rust/src/graph/higher_order_queries.rs

```diff
@@ -327,4 +327,57 @@ mod tests {
             ])
         );
     }
+
+    #[test]
+    fn test_generate_module_permutations_two_closed_layers() {
+        let mut graph = Graph::default();
+
+        // Create four layers with the middle two closed
+        let top_module = graph.get_or_add_module("app.top").token;
+        let middle_top_module = graph.get_or_add_module("app.middle-top").token;
+        let middle_bottom_module = graph.get_or_add_module("app.middle-bottom").token;
+        let bottom_module = graph.get_or_add_module("app.bottom").token;
+
+        let mut top_layer = FxHashSet::default();
+        top_layer.insert(top_module);
+
+        let mut middle_top_layer = FxHashSet::default();
+        middle_top_layer.insert(middle_top_module);
+
+        let mut middle_bottom_layer = FxHashSet::default();
+        middle_bottom_layer.insert(middle_bottom_module);
+
+        let mut bottom_layer = FxHashSet::default();
+        bottom_layer.insert(bottom_module);
+
+        let top_level = Level::new(top_layer, false, false);
+        let middle_top_level = Level::new(middle_top_layer, false, true); // Closed
+        let middle_bottom_level = Level::new(middle_bottom_layer, false, true); // Closed
```

Comment:
Good to have a test for two closed layers together, but feels like we could also do with something along these lines

```
one
two
three (closed)
four
five
six (closed)
seven
eight
```

I'd put that in the Python test though probably.

Reply:
I tested this case:

```
layers = [
    Layer("highest"),
    Layer("high", closed=True),
    Layer("mid"),
    Layer("low", closed=True),
    Layer("lowest"),
]
```

I'm not sure what benefit we'd get from testing even more layers, so kept it simple (in your example you put two open layers "four" and "five" between the closed layers) 🤔